tests/functional/fixtures/Dockerfile

```diff
@@ -0,0 +1,8 @@
+ARG GITLAB_IMAGE=get_this_from_env
+ARG GITLAB_TAG=get_this_from_env
+FROM ${GITLAB_IMAGE}:${GITLAB_TAG}
+
+COPY create_license.rb /
+RUN ruby /create_license.rb
+
+CMD ["/assets/wrapper"]
```

Comment:
This is very cool! I wonder if we could bypass the entire Docker build step by using the same approach as in our current `set_token.rb`.

That is, plug `create_license.rb` into our pytest setup dynamically as a fixture only consumed by EE tests.
- Instead of using `initial_license_file`, add a license via the console the same way we do the token now
- See for example https://docs.gitlab.com/ee/administration/troubleshooting/gitlab_rails_cheat_sheet.html#add-a-license-through-the-console (may need some rework.. seems like wants a key not a file)
- Our token fixture is done via the console: https://github.com/python-gitlab/python-gitlab/blob/main/tests/functional/conftest.py#L34-L50)

This might have some benefits:
- ability to completely skip this step if EE tests aren't run
- for people using `--reuse-containers` locally for faster tests (https://github.com/python-gitlab/python-gitlab/blob/main/CONTRIBUTING.rst#running-integration-tests) this may also help a lot to keep things fast
- wouldn't need to rebuild twice when running CLI/API functional tests and less things to go wrong, I hope ;)

This is all in my head and haven't tried it yet.. not sure if you've already tried this approach and it didn't work. Let me know what you think.

Reply:
I tried something like that at first but didn't have success. The issue is that as soon as the container is launched then it starts running all the startup stages. So would have to create the key, shutdown the running gitlab server, and then restart it. Which became ugly...

This is because we generate a public/private key pair which we need GitLab to use the public key. Then we also create the license key. So yes we could add the license key using the API but we have to have our own generated public key in place before we can add our license key.

Running the docker creation doesn't really take much time, like 10 seconds or so, assuming the gitlab image is already cached. And we have to retrieve the gitlab image either way.

Reply:
Ah ok, that's a shame, thanks a lot for the effort! In this case maybe we need to make a follow-up to update docs for `--reuse-containers` or I should check if it still works after this PR.

Reply:
The issue is that the public key is buried deep inside the directory tree at:
`/opt/gitlab/embedded/service/gitlab-rails/.license_encryption_key.pub`

So not sure what magic we can do to replace that. But yeah something for later 😊

Reply:
I think this is resolved at the moment.

Reply:
Thanks a lot again @JohnVillalovos for this! Did some research and I think I've found a solution that won't break the `--keep-containers` flag and with no docker build penalty.

This'll just inject it into the entrypoint at runtime before starting the wrapper. Could you try this patch? It works for me locally. It does revert some of the changes as they should no longer be needed with this. Also gives me ideas to further simplify our docker setup.. and if we ever move to GitLab, should be easier to do it [without compose](https://docs.gitlab.com/ee/ci/services/gitlab.html) in CI.

```diff
diff --git a/tests/functional/fixtures/.env b/tests/functional/fixtures/.env
index d7981d5..e7be6c9 100644
--- a/tests/functional/fixtures/.env
+++ b/tests/functional/fixtures/.env
@@ -1,5 +1,2 @@
 GITLAB_IMAGE=gitlab/gitlab-ee
 GITLAB_TAG=14.9.2-ee.0
-
-GITLAB_CI_IMAGE=localhost/gitlab-ci
-GITLAB_CI_TAG=latest
diff --git a/tests/functional/fixtures/build-dockerfile.sh b/tests/functional/fixtures/build-dockerfile.sh
deleted file mode 100755
index 5fa0381..0000000
--- a/tests/functional/fixtures/build-dockerfile.sh
+++ /dev/null
@@ -1,17 +0,0 @@
-#!/bin/bash
-
-set -u
-set -e
-
-TOP_DIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
-
-cd "${TOP_DIR}" || exit 1
-
-source .env
-
-docker build \
-    -t "${GITLAB_CI_IMAGE}:${GITLAB_CI_TAG}" \
-    --build-arg GITLAB_IMAGE="${GITLAB_IMAGE}" \
-    --build-arg GITLAB_TAG="${GITLAB_TAG}" \
-    --no-cache \
-    "${TOP_DIR}"
diff --git a/tests/functional/fixtures/docker-compose.yml b/tests/functional/fixtures/docker-compose.yml
index 921e0e6..9ad8922 100644
--- a/tests/functional/fixtures/docker-compose.yml
+++ b/tests/functional/fixtures/docker-compose.yml
@@ -6,7 +6,7 @@ networks:
 
 services:
   gitlab:
-    image: '${GITLAB_CI_IMAGE}:${GITLAB_CI_TAG}'
+    image: '${GITLAB_IMAGE}:${GITLAB_TAG}'
     container_name: 'gitlab-test'
     hostname: 'gitlab.test'
     privileged: true # Just in case https://gitlab.com/gitlab-org/omnibus-gitlab/-/issues/1350
@@ -31,7 +31,10 @@ services:
         gitlab_exporter['enable'] = false
         grafana['enable'] = false
         letsencrypt['enable'] = false
-        gitlab_rails['initial_license_file'] = "/python-gitlab-ci.gitlab-license"
+        gitlab_rails['initial_license_file'] = '/python-gitlab-ci.gitlab-license'
+    entrypoint: ['/bin/sh', '-c', 'ruby /create_license.rb && /assets/wrapper']
+    volumes:
+      - ${PWD}/tests/functional/fixtures/create_license.rb:/create_license.rb
     ports:
       - '8080:80'
       - '2222:22'
diff --git a/tox.ini b/tox.ini
index 2c2cfe4..aafef8e 100644
--- a/tox.ini
+++ b/tox.ini
@@ -5,7 +5,7 @@ skip_missing_interpreters = True
 envlist = py310,py39,py38,py37,flake8,black,twine-check,mypy,isort,cz,pylint
 
 [testenv]
-passenv = GITLAB_IMAGE GITLAB_TAG PY_COLORS NO_COLOR FORCE_COLOR DOCKER_HOST
+passenv = GITLAB_IMAGE GITLAB_TAG PY_COLORS NO_COLOR FORCE_COLOR DOCKER_HOST PWD
 setenv = VIRTUAL_ENV={envdir}
 whitelist_externals = true
 usedevelop = True
@@ -104,13 +104,11 @@ exclude_lines =
 [testenv:cli_func_v4]
 deps = -r{toxinidir}/requirements-docker.txt
 commands =
-  {toxinidir}/tests/functional/fixtures/build-dockerfile.sh
   pytest --script-launch-mode=subprocess --cov --cov-report xml tests/functional/cli {posargs}
 
 [testenv:api_func_v4]
 deps = -r{toxinidir}/requirements-docker.txt
 commands =
-  {toxinidir}/tests/functional/fixtures/build-dockerfile.sh
   pytest --cov --cov-report xml tests/functional/api {posargs}
 
 [testenv:smoke]
```

Reply:
Done. Thanks for the info @nejch